cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+
+	"github.com/golang/glog"
+	gce "google.golang.org/api/compute/v1"
+)
+
+// GceCache is used for caching cluster resources state.
+type GceCache struct {
+	// Cache content.
+	migs            []*MigInformation
+	instancesCache  map[GceRef]Mig
+	resourceLimiter *cloudprovider.ResourceLimiter
+	machinesCache   map[MachineTypeKey]*gce.MachineType
+	// Locks. Rules of locking:
+	// - migsMutex protects only migs.
+	// - cacheMutex protects instancesCache, resourceLimiter and machinesCache.
+	// - if both locks are needed, cacheMutex must be obtained before migsMutex.
+	cacheMutex sync.Mutex
+	migsMutex  sync.Mutex
+	// Service used to refresh cache.
+	GceService AutoscalingGceClient
+}
+
+// MigInformation is a wrapper for Mig.
+type MigInformation struct {
+	Config   Mig
+	Basename string
+}
+
+// MachineTypeKey is used to identify MachineType.
+type MachineTypeKey struct {
+	Zone        string
+	MachineType string
+}
+
+// NewGceCache creates empty GceCache.
+func NewGceCache(gceService AutoscalingGceClient) GceCache {
+	return GceCache{
+		migs:           []*MigInformation{},
+		instancesCache: map[GceRef]Mig{},
+		machinesCache:  map[MachineTypeKey]*gce.MachineType{},
+		GceService:     gceService,
+	}
+}
+
+//  Methods locking on migsMutex.
+
+// RegisterMig returns true if the node group wasn't in cache before, or its config was updated.
+func (gc *GceCache) RegisterMig(mig Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for i := range gc.migs {
+		if oldMig := gc.migs[i].Config; oldMig.GceRef() == mig.GceRef() {
+			if !reflect.DeepEqual(oldMig, mig) {
+				gc.migs[i].Config = mig
+				glog.V(4).Infof("Updated Mig %s", mig.GceRef().String())
+				return true
+			}
+			return false
+		}
+	}
+
+	glog.V(1).Infof("Registering %s", mig.GceRef().String())
+	gc.migs = append(gc.migs, &MigInformation{
+		Config: mig,
```

Comment:
Can you add TODO: fetch and set Basename here? Currently we rely on someone outside of cache calling RegenerateInstancesCache before GetMigForInstance (if they called Mig with empty basename matches the prefix of every node bad things will happen). At least that's the impression I get from the code.

Reply:
Done